File: src/main/java/dev/jolvera/finjav/controllers/MainPageController.java
package dev.jolvera.finjav.controllers;

import dev.jolvera.finjav.FinJavComponent;
import dev.jolvera.finjav.HelloApplication;
import dev.jolvera.finjav.viewModels.MainViewModel;
import jakarta.inject.Inject;
import javafx.application.Platform;
import javafx.fxml.FXML;
import javafx.fxml.FXMLLoader;
import javafx.scene.control.*;

import java.util.Optional;

public class MainPageController {
    @FXML private TextField SearchField;
    @FXML private Button AddRecurrenceButton;
    @FXML private Button RefreshButton;
    @FXML private MenuItem AccountMenuButton;
    @FXML private Label CurrentUserDisplay;

    private MainViewModel viewModel;
    private FinJavComponent component;

    @Inject
    public MainPageController(MainViewModel viewModel, FinJavComponent component) {
        this.viewModel = viewModel;
        this.component = component;
    }

    @FXML
    private void initialize() {
        setupHandlers();
        setupListeners();

        if (viewModel.activeUserProperty().getValue() == null) {
            Platform.runLater(this::handleOpenAccountDialog);
        }
    }

    private void setupHandlers() {
        AddRecurrenceButton.setOnAction(e -> System.out.println("addRecurrence pressed"));
        RefreshButton.setOnAction(e -> System.out.println("refresh pressed"));
    }

    private void setupListeners() {
        viewModel.activeUserProperty().addListener((observable, oldValue, newValue) -> {
            if (newValue == null) {
                CurrentUserDisplay.setText("");
                AccountMenuButton.setText("Log in");
                AccountMenuButton.setOnAction(e -> handleOpenAccountDialog());
                handleOpenAccountDialog();
            } else {
                CurrentUserDisplay.setText("User: " + newValue.getName());
                AccountMenuButton.setText("Log out");
                AccountMenuButton.setOnAction(e -> logOut());
            }
        });
    }

    private void handleOpenAccountDialog() {
        try {
            FXMLLoader loader = new FXMLLoader(HelloApplication.class.getResource("account-dialog.fxml"));
            AccountDialogController controller = component.accountDialogController();
            loader.setController(controller);
            DialogPane dialogPane = loader.load();

            Dialog<User> dialog = new Dialog<>();
            dialog.setDialogPane(dialogPane);
            dialog.setTitle("Login/Register Account");

            Optional<User> result = dialog.showAndWait();

            if (result.isPresent()) {
                User user = result.get();
                viewModel.activeUserProperty().setValue(user);
            } else {
                User controllerResult = controller.getResult();
                if (controllerResult != null) {
                    viewModel.activeUserProperty().setValue(controllerResult);
                } else {
                    System.out.println("Account dialog was cancelled");
                }
            }
        } catch  (Exception ex) {
            System.out.println(ex.getMessage());
        }
    }

    private void logOut() {
        viewModel.activeUserProperty().setValue(null);
    }

}

File: src/main/java/dev/jolvera/finjav/controllers/MainPageController.kt
package dev.jolvera.finjav.controllers

import dev.jolvera.finjav.viewModels.MainViewModel
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.fxml.FXML
import javafx.scene.control.Button
import javafx.scene.control.Label
import javafx.scene.control.MenuItem
import javafx.scene.control.TextField

class MainPageController(private val viewModel: MainViewModel) {
    @FXML private val SearchField: TextField? = null
    @FXML private val AddRecurrenceButton: Button? = null
    @FXML private val RefreshButton: Button? = null
    @FXML private val AccountMenuButton: MenuItem? = null
    @FXML private val CurrentUserDisplay: Label? = null

    @FXML
    private fun initialize() {
        setupHandlers()
        setupListeners()
    }

    private fun setupListeners() {
        viewModel.activeUserProperty().addListener { _, _, newUser ->
            if (newUser == null) {
                CurrentUserDisplay!!.text = "No User"
                AccountMenuButton!!.text = "Log In"
                AccountMenuButton!!.onAction = EventHandler { e -> handleOpenAccountDialog() }
                handleOpenAccountDialog()
            } else {
                CurrentUserDisplay!!.text = "User: " + newUser.name
                AccountMenuButton!!.text = "Log out"
                AccountMenuButton.onAction = EventHandler { e: ActionEvent? -> logOut() }
            }
        }
    }

    private fun setupHandlers() {
        AddRecurrenceButton!!.onAction = EventHandler { e -> println("AddRecurrenceButton pressed!") }
        RefreshButton!!.onAction = EventHandler { e -> println("RefreshButton pressed!") }
    }

    private fun logOut() { viewModel.activeUserProperty().set(null) }

    private fun handleOpenAccountDialog() {

    }
}